iot_packet_verifier/src/pending.rs

```diff
@@ -0,0 +1,467 @@
+use async_trait::async_trait;
+use chrono::{DateTime, Duration, Utc};
+use helium_crypto::PublicKeyBinary;
+use solana::SolanaNetwork;
+use solana_sdk::signature::Signature;
+use sqlx::{postgres::PgRow, FromRow, PgPool, Postgres, Row, Transaction};
+use std::{collections::HashMap, sync::Arc};
+use tokio::sync::Mutex;
+
+/// To avoid excessive burn transaction (which cost us money), we institute a minimum
+/// amount of Data Credits accounted for before we burn from a payer:
+const BURN_THRESHOLD: i64 = 10_000;
+
+#[async_trait]
+pub trait AddPendingBurn {
+    async fn add_burned_amount(
+        &mut self,
+        payer: &PublicKeyBinary,
+        amount: u64,
+    ) -> Result<(), sqlx::Error>;
+}
+
+#[async_trait]
+pub trait PendingTables {
+    type Transaction<'a>: PendingTablesTransaction<'a> + Send + Sync
+    where
+        Self: 'a;
+
+    async fn fetch_next_burn(&self) -> Result<Option<Burn>, sqlx::Error>;
+
+    async fn fetch_all_pending_burns(&self) -> Result<Vec<Burn>, sqlx::Error>;
+
+    async fn fetch_all_pending_txns(&self) -> Result<Vec<PendingTxn>, sqlx::Error>;
+
+    async fn submit_txn(
+        &self,
+        payer: &PublicKeyBinary,
+        amount: u64,
+        signature: &Signature,
+    ) -> Result<(), sqlx::Error>;
+
+    async fn begin<'a>(&'a self) -> Result<Self::Transaction<'a>, sqlx::Error>;
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ConfirmPendingError<S> {
+    #[error("Sqlx error: {0}")]
+    SqlxError(#[from] sqlx::Error),
+    #[error("Chrono error: {0}")]
+    ChronoError(#[from] chrono::OutOfRangeError),
+    #[error("Solana error: {0}")]
+    SolanaError(S),
+}
+
+pub async fn confirm_pending_txns<S>(
+    pending_tables: &impl PendingTables,
+    solana: &S,
+) -> Result<(), ConfirmPendingError<S::Error>>
+where
+    S: SolanaNetwork,
+{
+    // Fetch all pending transactions and confirm them
+    let pending = pending_tables.fetch_all_pending_txns().await?;
+    for pending in pending {
+        // Sleep for at least a minute since the time of submission to
+        // give the transaction plenty of time to be confirmed
+        let time_since_submission = Utc::now() - pending.time_of_submission;
+        if Duration::minutes(1) > time_since_submission {
+            tokio::time::sleep((Duration::minutes(1) - time_since_submission).to_std()?).await;
+        }
+
+        let mut txn = pending_tables.begin().await?;
+        // We remove the transaction regardless of whether it has been confirmed
+        // or not:
+        txn.confirm_txn(&pending.signature).await?;
```

Comment:
what does it mean to "remove the transaction regardless of whether or not it's been confirmed" if we only remove the the burn amount from the pending burns table if it _is_ confirmed? what are we removing regardless?

Reply:
I explained this comment a little further, but I think the renamed methods should also help quite significantly here